Firestore/core/src/firebase/firestore/remote/remote_objc_bridge.mm

```diff
@@ -127,170 +114,232 @@ bool IsLoggingEnabled() {
-  request.database = [serializer_ encodedDatabaseID];
-  request.addTarget = [serializer_ encodedTarget:query];
-  request.labels = [serializer_ encodedListenRequestLabelsForQueryData:query];
-  return request;
+WatchStreamSerializer::WatchStreamSerializer(Serializer serializer)
+    : serializer_{std::move(serializer)} {
 }
 
-GCFSListenRequest* WatchStreamSerializer::CreateUnwatchRequest(
-    TargetId target_id) const {
-  GCFSListenRequest* request = [GCFSListenRequest message];
-  request.database = [serializer_ encodedDatabaseID];
-  request.removeTarget = target_id;
-  return request;
+Message<google_firestore_v1_ListenRequest>
+WatchStreamSerializer::CreateWatchRequest(const QueryData& query) const {
+  Message<google_firestore_v1_ListenRequest> result{
+      google_firestore_v1_ListenRequest_fields};
+  auto& request = result.mutable_proto();
+
+  request.database = serializer_.EncodeDatabaseId();
+  request.which_target_change =
+      google_firestore_v1_ListenRequest_add_target_tag;
+  request.add_target = serializer_.EncodeTarget(query);
+
+  auto labels = serializer_.EncodeListenRequestLabels(query);
+  if (!labels.empty()) {
+    request.labels_count = nanopb::CheckedSize(labels.size());
+    request.labels = MakeArray<google_firestore_v1_ListenRequest_LabelsEntry>(
+        request.labels_count);
+
+    pb_size_t i = 0;
+    for (const auto& kv : labels) {
+      request.labels[i].key = Serializer::EncodeString(kv.first);
+      request.labels[i].value = Serializer::EncodeString(kv.second);
+      ++i;
+    }
+  }
+
+  return result;
 }
 
-grpc::ByteBuffer WatchStreamSerializer::ToByteBuffer(
-    GCFSListenRequest* request) {
-  return ConvertToByteBuffer([request data]);
+Message<google_firestore_v1_ListenRequest>
+WatchStreamSerializer::CreateUnwatchRequest(TargetId target_id) const {
+  Message<google_firestore_v1_ListenRequest> result{
+      google_firestore_v1_ListenRequest_fields};
+  auto& request = result.mutable_proto();
+
+  request.database = serializer_.EncodeDatabaseId();
+  request.which_target_change =
+      google_firestore_v1_ListenRequest_remove_target_tag;
+  request.remove_target = target_id;
+
+  return result;
 }
 
-GCFSListenResponse* WatchStreamSerializer::ParseResponse(
-    const grpc::ByteBuffer& message, Status* out_status) const {
-  return ToProto<GCFSListenResponse>(message, out_status);
+MaybeMessage<google_firestore_v1_ListenResponse>
+WatchStreamSerializer::ParseResponse(const grpc::ByteBuffer& message) const {
+  return Message<google_firestore_v1_ListenResponse>::Parse(
+      google_firestore_v1_ListenResponse_fields, message);
 }
 
 std::unique_ptr<WatchChange> WatchStreamSerializer::ToWatchChange(
-    GCFSListenResponse* proto) const {
-  return [serializer_ decodedWatchChange:proto];
+    const google_firestore_v1_ListenResponse& response) const {
+  nanopb::Reader reader;
+  return serializer_.DecodeWatchChange(&reader, response);
```

Comment:
This ignores the failure that might happen while decoding.

Reply:
Postponing for now; let's discuss error handling in the thread below.